File: mm-cs/src/MetaMessage/Jsonc/ValueType.cs
namespace MetaMessage.Jsonc;

public enum ValueType
{
    Unknown,

    Doc,
    Slice,
    Array,
    Struct,
    Map,

    String,
    Bytes,
    Bool,

    Int,
    Int8,
    Int16,
    Int32,
    Int64,
    Uint,
    Uint8,
    Uint16,
    Uint32,
    Uint64,

    Float32,
    Float64,

    BigInt,
    DateTime,
    Date,
    Time,

    UUID,
    Decimal,
    IP,
    URL,
    Email,

    Enum,

    Image,
    Video
}

public static class ValueTypeConstants
{
    public const string UnknownStr = "unknown";
    public const string DocStr = "doc";
    public const string ArrayStr = "arr";
    public const string SliceStr = "slice";
    public const string ObjectStr = "obj";
    public const string MapStr = "map";
    public const string StringStr = "str";
    public const string BytesStr = "bytes";
    public const string BoolStr = "bool";
    public const string IntStr = "i";
    public const string Int8Str = "i8";
    public const string Int16Str = "i16";
    public const string Int32Str = "i32";
    public const string Int64Str = "i64";
    public const string UintStr = "u";
    public const string Uint8Str = "u8";
    public const string Uint16Str = "u16";
    public const string Uint32Str = "u32";
    public const string Uint64Str = "u64";
    public const string Float32Str = "f32";
    public const string Float64Str = "f64";
    public const string BigIntStr = "bi";
    public const string DateTimeStr = "datetime";
    public const string DateStr = "date";
    public const string TimeStr = "time";
    public const string UUIDStr = "uuid";
    public const string DecimalStr = "decimal";
    public const string IPStr = "ip";
    public const string URLStr = "url";
    public const string EmailStr = "email";
    public const string EnumStr = "enum";
    public const string ImageStr = "image";
    public const string VideoStr = "video";
}

public static class ValueTypeExtensions
{
    private static readonly Dictionary<string, ValueType> StringToValueType = new()
    {
        { ValueTypeConstants.UnknownStr, ValueType.Unknown },
        { ValueTypeConstants.DocStr, ValueType.Doc },
        { ValueTypeConstants.ArrayStr, ValueType.Array },
        { ValueTypeConstants.SliceStr, ValueType.Slice },
        { ValueTypeConstants.ObjectStr, ValueType.Struct },
        { ValueTypeConstants.MapStr, ValueType.Map },
        { ValueTypeConstants.StringStr, ValueType.String },
        { ValueTypeConstants.BytesStr, ValueType.Bytes },
        { ValueTypeConstants.BoolStr, ValueType.Bool },
        { ValueTypeConstants.IntStr, ValueType.Int },
        { ValueTypeConstants.Int8Str, ValueType.Int8 },
        { ValueTypeConstants.Int16Str, ValueType.Int16 },
        { ValueTypeConstants.Int32Str, ValueType.Int32 },
        { ValueTypeConstants.Int64Str, ValueType.Int64 },
        { ValueTypeConstants.UintStr, ValueType.Uint },
        { ValueTypeConstants.Uint8Str, ValueType.Uint8 },
        { ValueTypeConstants.Uint16Str, ValueType.Uint16 },
        { ValueTypeConstants.Uint32Str, ValueType.Uint32 },
        { ValueTypeConstants.Uint64Str, ValueType.Uint64 },
        { ValueTypeConstants.Float32Str, ValueType.Float32 },
        { ValueTypeConstants.Float64Str, ValueType.Float64 },
        { ValueTypeConstants.BigIntStr, ValueType.BigInt },
        { ValueTypeConstants.DateTimeStr, ValueType.DateTime },
        { ValueTypeConstants.DateStr, ValueType.Date },
        { ValueTypeConstants.TimeStr, ValueType.Time },
        { ValueTypeConstants.UUIDStr, ValueType.UUID },
        { ValueTypeConstants.DecimalStr, ValueType.Decimal },
        { ValueTypeConstants.IPStr, ValueType.IP },
        { ValueTypeConstants.URLStr, ValueType.URL },
        { ValueTypeConstants.EmailStr, ValueType.Email },
        { ValueTypeConstants.EnumStr, ValueType.Enum },
        { ValueTypeConstants.ImageStr, ValueType.Image },
        { ValueTypeConstants.VideoStr, ValueType.Video }
    };

    public static string ToTypeString(this ValueType vt)
    {
        return vt switch
        {
            ValueType.Unknown => ValueTypeConstants.UnknownStr,
            ValueType.Doc => ValueTypeConstants.DocStr,
            ValueType.Array => ValueTypeConstants.ArrayStr,
            ValueType.Slice => ValueTypeConstants.SliceStr,
            ValueType.Struct => ValueTypeConstants.ObjectStr,
            ValueType.Map => ValueTypeConstants.MapStr,
            ValueType.String => ValueTypeConstants.StringStr,
            ValueType.Bytes => ValueTypeConstants.BytesStr,
            ValueType.Bool => ValueTypeConstants.BoolStr,
            ValueType.Int => ValueTypeConstants.IntStr,
            ValueType.Int8 => ValueTypeConstants.Int8Str,
            ValueType.Int16 => ValueTypeConstants.Int16Str,
            ValueType.Int32 => ValueTypeConstants.Int32Str,
            ValueType.Int64 => ValueTypeConstants.Int64Str,
            ValueType.Uint => ValueTypeConstants.UintStr,
            ValueType.Uint8 => ValueTypeConstants.Uint8Str,
            ValueType.Uint16 => ValueTypeConstants.Uint16Str,
            ValueType.Uint32 => ValueTypeConstants.Uint32Str,
            ValueType.Uint64 => ValueTypeConstants.Uint64Str,
            ValueType.Float32 => ValueTypeConstants.Float32Str,
            ValueType.Float64 => ValueTypeConstants.Float64Str,
            ValueType.BigInt => ValueTypeConstants.BigIntStr,
            ValueType.DateTime => ValueTypeConstants.DateTimeStr,
            ValueType.Date => ValueTypeConstants.DateStr,
            ValueType.Time => ValueTypeConstants.TimeStr,
            ValueType.UUID => ValueTypeConstants.UUIDStr,
            ValueType.Decimal => ValueTypeConstants.DecimalStr,
            ValueType.IP => ValueTypeConstants.IPStr,
            ValueType.URL => ValueTypeConstants.URLStr,
            ValueType.Email => ValueTypeConstants.EmailStr,
            ValueType.Enum => ValueTypeConstants.EnumStr,
            ValueType.Image => ValueTypeConstants.ImageStr,
            ValueType.Video => ValueTypeConstants.VideoStr,
            _ => ValueTypeConstants.UnknownStr
        };
    }

    public static ValueType ParseValueType(string s)
    {
        s = s.ToLower();
        if (StringToValueType.TryGetValue(s, out var vt))
        {
            return vt;
        }
        return ValueType.Unknown;
    }

    public static bool NeedsQuotes(this ValueType vt)
    {
        return vt switch
        {
            ValueType.String => true,
            ValueType.Bytes => true,
            ValueType.DateTime => true,
            ValueType.Date => true,
            ValueType.Time => true,
            ValueType.UUID => true,
            ValueType.IP => true,
            ValueType.URL => true,
            ValueType.Email => true,
            ValueType.Enum => true,
            _ => false
        };
    }
}

File: mm-rs/src/ir/value_type.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ValueType {
    Unknown,
    Doc,
    Vec,
    Arr,
    Obj,
    Map,
    Str,
    Bytes,
    Bool,
    I,
    I8,
    I16,
    I32,
    I64,
    U,
    U8,
    U16,
    U32,
    U64,
    F32,
    F64,
    Bigint,
    Datetime,
    Date,
    Time,
    Uuid,
    Decimal,
    Ip,
    Url,
    Email,
    Enum,
    Image,
    Video,
}

impl ValueType {
    #[allow(clippy::should_implement_trait)]
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "unknown" => ValueType::Unknown,
            "doc" => ValueType::Doc,
            "vec" => ValueType::Vec,
            "arr" => ValueType::Arr,
            "obj" => ValueType::Obj,
            "map" => ValueType::Map,
            "str" => ValueType::Str,
            "bytes" => ValueType::Bytes,
            "bool" => ValueType::Bool,
            "i" => ValueType::I,
            "i8" => ValueType::I8,
            "i16" => ValueType::I16,
            "i32" => ValueType::I32,
            "i64" => ValueType::I64,
            "u" => ValueType::U,
            "u8" => ValueType::U8,
            "u16" => ValueType::U16,
            "u32" => ValueType::U32,
            "u64" => ValueType::U64,
            "f32" => ValueType::F32,
            "f64" => ValueType::F64,
            "bigint" => ValueType::Bigint,
            "datetime" => ValueType::Datetime,
            "date" => ValueType::Date,
            "time" => ValueType::Time,
            "uuid" => ValueType::Uuid,
            "decimal" => ValueType::Decimal,
            "ip" => ValueType::Ip,
            "url" => ValueType::Url,
            "email" => ValueType::Email,
            "enum" => ValueType::Enum,
            "image" => ValueType::Image,
            "video" => ValueType::Video,
            _ => ValueType::Unknown,
        }
    }

    pub fn to_str(&self) -> &str {
        match self {
            ValueType::Unknown => "unknown",
            ValueType::Doc => "doc",
            ValueType::Vec => "vec",
            ValueType::Arr => "arr",
            ValueType::Obj => "obj",
            ValueType::Map => "map",
            ValueType::Str => "str",
            ValueType::Bytes => "bytes",
            ValueType::Bool => "bool",
            ValueType::I => "i",
            ValueType::I8 => "i8",
            ValueType::I16 => "i16",
            ValueType::I32 => "i32",
            ValueType::I64 => "i64",
            ValueType::U => "u",
            ValueType::U8 => "u8",
            ValueType::U16 => "u16",
            ValueType::U32 => "u32",
            ValueType::U64 => "u64",
            ValueType::F32 => "f32",
            ValueType::F64 => "f64",
            ValueType::Bigint => "bigint",
            ValueType::Datetime => "datetime",
            ValueType::Date => "date",
            ValueType::Time => "time",
            ValueType::Uuid => "uuid",
            ValueType::Decimal => "decimal",
            ValueType::Ip => "ip",
            ValueType::Url => "url",
            ValueType::Email => "email",
            ValueType::Enum => "enum",
            ValueType::Image => "image",
            ValueType::Video => "video",
        }
    }

    pub fn needs_quotes(&self) -> bool {
        matches!(
            self,
            ValueType::Unknown
                | ValueType::Str
                | ValueType::Bytes
                | ValueType::Datetime
                | ValueType::Date
                | ValueType::Time
                | ValueType::Uuid
                | ValueType::Ip
                | ValueType::Url
                | ValueType::Email
                | ValueType::Enum
        )
    }

    pub fn from_code(code: u8) -> Self {
        match code {
            0 => ValueType::Unknown,
            1 => ValueType::Doc,
            2 => ValueType::Vec,
            3 => ValueType::Arr,
            4 => ValueType::Obj,
            5 => ValueType::Map,
            6 => ValueType::Str,
            7 => ValueType::Bytes,
            8 => ValueType::Bool,
            9 => ValueType::I,
            10 => ValueType::I8,
            11 => ValueType::I16,
            12 => ValueType::I32,
            13 => ValueType::I64,
            14 => ValueType::U,
            15 => ValueType::U8,
            16 => ValueType::U16,
            17 => ValueType::U32,
            18 => ValueType::U64,
            19 => ValueType::F32,
            20 => ValueType::F64,
            21 => ValueType::Bigint,
            22 => ValueType::Datetime,
            23 => ValueType::Date,
            24 => ValueType::Time,
            25 => ValueType::Uuid,
            26 => ValueType::Decimal,
            27 => ValueType::Ip,
            28 => ValueType::Url,
            29 => ValueType::Email,
            30 => ValueType::Enum,
            31 => ValueType::Image,
            32 => ValueType::Video,
            _ => ValueType::Unknown,
        }
    }
}
